server/src/test/java/org/opensearch/index/compositeindex/datacube/startree/builder/AbstractStarTreeBuilderTests.java

```diff
@@ -1278,7 +1598,166 @@ public void test_build_starTreeDataset() throws IOException {
+
+        IndexInput starTreeIndexInput = dataIn.slice(
+            "star-tree data slice for respective star-tree fields",
+            resultStarTreeMetadata.getDataStartFilePointer(),
+            resultStarTreeMetadata.getDataLength()
+        );
+        StarTree starTree = new StarTree(starTreeIndexInput, resultStarTreeMetadata);
+
+        StarTreeNode starTreeNode = starTree.getRoot();
+        Queue<StarTreeNode> expectedTreeNodeQueue = new ArrayDeque<>();
+        Queue<InMemoryTreeNode> resultTreeNodeQueue = new ArrayDeque<>();
+
+        expectedTreeNodeQueue.add(starTreeNode);
+        resultTreeNodeQueue.add(rootNode);
+
+        while ((starTreeNode = expectedTreeNodeQueue.poll()) != null && (rootNode = resultTreeNodeQueue.poll()) != null) {
+
+            // verify the star node
+            assertStarTreeNode(starTreeNode, rootNode);
+
+            Iterator<? extends StarTreeNode> expectedChildrenIterator = starTreeNode.getChildrenIterator();
+
+            List<InMemoryTreeNode> sortedChildren = new ArrayList<>();
+            if (rootNode.children != null) {
+                sortedChildren = new ArrayList<>(rootNode.children.values());
+            }
+            sortedChildren.sort(
+                Comparator.comparingInt(InMemoryTreeNode::getNodeType).thenComparingLong(InMemoryTreeNode::getDimensionValue)
+            );
+
+            if (starTreeNode.getChildDimensionId() != -1) {
+                assertFalse(sortedChildren.isEmpty());
+                int childCount = 0;
+                while (expectedChildrenIterator.hasNext()) {
+                    StarTreeNode child = expectedChildrenIterator.next();
+                    InMemoryTreeNode resultChildNode = sortedChildren.get(childCount);
+
+                    assertNotNull(resultChildNode);
+                    assertNotNull(child);
+                    assertStarTreeNode(child, resultChildNode);
+
+                    expectedTreeNodeQueue.add(child);
+                    resultTreeNodeQueue.add(resultChildNode);
+
+                    childCount++;
+                }
+                assertEquals(childCount, rootNode.children.size());
+            } else {
+                assertNull(rootNode.children);
+            }
+        }
+
+        assertTrue(expectedTreeNodeQueue.isEmpty());
+        assertTrue(resultTreeNodeQueue.isEmpty());
+
+    }
+
+    private void assertStarTreeNode(StarTreeNode starTreeNode, InMemoryTreeNode treeNode) throws IOException {
+        assertEquals(starTreeNode.getDimensionId(), treeNode.dimensionId);
+        assertEquals(starTreeNode.getDimensionValue(), treeNode.dimensionValue);
+        assertEquals(starTreeNode.getStartDocId(), treeNode.startDocId);
+        assertEquals(starTreeNode.getEndDocId(), treeNode.endDocId);
+        assertEquals(starTreeNode.getChildDimensionId(), treeNode.childDimensionId);
+        assertEquals(starTreeNode.getAggregatedDocId(), treeNode.aggregatedDocId);
```

Comment:
Lets get the actual docs from file formats and perform assertions against the build one ?

Reply:
```
    StarTreeDocument[] getSegmentsStarTreeDocuments(List<StarTreeValues> starTreeValuesSubs) throws IOException {
```
Lets check out how to reuse merge logic which forms the star tree documents